MAINT/ENH: resolve mypy warnings/errors

mckib2 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

Companion PR to gh-13313

#### What does this implement/fix?
<!--Please explain your changes.-->

- resolves all current `mypy` warnings/errors
    - some modules added to ignore list in `mypy.ini`
    - adds types when necessary in py files
    - adds a couple stub files for pyx files

#### Additional information
<!--Any additional information you think is important.-->

- To work with `numpy<1.20`, `numpy-stubs` needs to be installed
- Might think about running `python runtests.py --mypy` during CI to keep an eye on it (doesn't need to cause failure, but good to know when things start breaking)

mckib2 commented:
@mdhaber Something seems to be going sideways with linprog docs somehow due to the changes in this PR.  Is there an easy answer here that I'm not seeing?  I'll attach the log, can provide more verbose output if needed.
[sphinx-err-v6t_28sg.log](https://github.com/scipy/scipy/files/5761167/sphinx-err-v6t_28sg.log)

person142 commented:
> Some typing features are not available for Pythnon 3.7 so they are simply ignored using a # type: ignore directive (see python/mypy#2087 (comment) for a workaround that uses typing.Protocol which is only available in Python 3.8 or through the typing-extensions module)

Use `typing_extensions`; we expect it to be available if anyone is using types with SciPy. (But guard it in an `if TYPE_CHECKING` block to not make it a runtime dependency.)

mckib2 commented:
I added some `from __future__ import annotations` and that seems to have cleared up the docs issues (fingers crossed).  Also added back in static typing dependent on `typing_extensions`.  @person142 Do you mind doing another once over to make sure everything looks okay?

person142 commented on scipy/integrate/_quadrature.py:
```diff
@@ -47,6 +49,20 @@ class AccuracyWarning(Warning):
+if TYPE_CHECKING:
+    # workaround for mypy function attributes see:
+    # https://github.com/python/mypy/issues/2087#issuecomment-462726600
+    from typing import Protocol, Dict, Tuple, Any
```
So I'd import everything except `Protocol` from typing, and then import `Protocol` from `typing_extensions`. Then instead of inlining `cache_decorator` here I'd probably do

```
if TYPE_CHECKING:
    class CacheAttributes(Protocol):
        cache: Dict[int, Tuple[Any, Any]]
else:
    CacheAttributes = Any


def cache_decorator(func: Callable) -> CacheAttributes:
    ...
```

mckib2 replied:
Done

person142 commented on scipy/integrate/_quadrature.py:
```diff
@@ -47,6 +49,20 @@ class AccuracyWarning(Warning):
+    from typing import Protocol, Dict, Tuple, Any
+    class CacheAttributes(Protocol):
+        cache: Dict[int, Tuple[Any, Any]]
+    def cache_decorator(func: Any) -> CacheAttributes:
```
Probably want to make `func` at least `Callable`.

mckib2 replied:
Done

person142 commented on scipy/optimize/_trustregion_constr/report.py:
```diff
@@ -1,10 +1,12 @@
-    COLUMN_NAMES = NotImplemented
-    COLUMN_WIDTHS = NotImplemented
-    ITERATION_FORMATS = NotImplemented
+    COLUMN_NAMES: List[str] = NotImplemented
```
Hm... the type here definitely doesn't match `NotImplemented`, but I'm also not sure what's up with them being set to `NotImplemented`. Was this supposed to be an ABC or something?

mckib2 replied:
It looks to me like it was intended to be an ABC.  Adding types here seems to make mypy do "the right thing", i.e. identify types correctly classes that inherit

mckib2 commented on scipy/optimize/__nnls.pyi:
```diff
@@ -0,0 +1,11 @@
+if TYPE_CHECKING:
+    import numpy.typing as npt
+
+def nnls(a: npt.ArrayLike, mda: int, m: int, n: int,
```
Done

person142 reviewed (approved):
(Would merge but CI isn't done yet.)
